R/ModuleInstantiation.R

```diff
@@ -73,7 +77,231 @@ ensureAllModulesInstantiated <- function(analysisSpecifications) {
+    if (!silent) {
+      message("ERROR - renv.lock file missing.")
+    }
+    return(
+      verifyModuleInstallationReturnValue(
+        moduleFolder = moduleFolder,
+        moduleInstalled = FALSE
+      )
+    )
+  }
+
+  # Check to see if we've already performed the verification by looking at the
+  # moduleStatus.txt file to see if the md5 in that file matches the one
+  # created by hashing the renv.lock file
+  lockfileContents <- ParallelLogger::loadSettingsFromJson(
+    fileName = file.path(moduleFolder, renvLockFileName)
+  )
+  lockfileHash <- digest::digest(
+    object = lockfileContents,
+    algo = "md5"
+  )
+  if (!forceVerification && file.exists(file.path(moduleFolder, moduleStatusFileName))) {
+    lockfileHashFromModuleStatusFile <- SqlRender::readSql(
+      sourceFile = file.path(moduleFolder, moduleStatusFileName)
+    )
+
+    # If the values match, the module is installed correctly
+    # return and exit
+    if (lockfileHashFromModuleStatusFile == lockfileHash) {
+      if (!silent) {
+        message("MODULE READY!")
+      }
+      return(
+        verifyModuleInstallationReturnValue(
+          moduleFolder = moduleFolder,
+          moduleInstalled = TRUE
+          )
+      )
+    }
+  }
+
+
+  # Now perform the consistency check to verify that the renv::restore()
+  # process executed successfully. We must do this in the module's context
+  Strategus:::withModuleRenv(
+    code = {
+      # Start by turning off verbose output to hide renv output
+      verboseOption <- getOption("renv.verbose")
+      options(renv.verbose = FALSE)
+      on.exit(options(renv.verbose = verboseOption))
+
+      # Get the renv project status and then identify the packages used
+      # in the project to determine if there were issues when restoring
+      # the project from the renv.lock file.
+      projectStatus <- renv::status()
+
+      # Get the packages in the project - adapted from
+      # https://github.com/rstudio/renv/blob/v1.0.3/R/status.R
+      project <- renv:::renv_project_resolve()
+      libpaths <- renv:::renv_libpaths_resolve()
+      dependencies <- renv:::renv_snapshot_dependencies(project, dev = FALSE)
+      packages <- sort(union(dependencies, "renv"))
+      paths <- renv:::renv_package_dependencies(packages, libpaths = libpaths, project = project)
+      packages <- as.character(names(paths))
+      # remove ignored packages
+      ignored <- c(
+        renv:::renv_project_ignored_packages(project),
+        renv:::renv_packages_base()
+      )
+      packages <- setdiff(packages, ignored)
+      projectStatus$packages <- packages
+      saveRDS(projectStatus, file="projectStatus.rds")
+    },
+    moduleFolder = moduleFolder
+  )
+
+  # The module's project status is written to the
+  # file system. Now we can get the module status and use the information
+  # to determine the restoration status
+  projectStatus <- readRDS(file.path(moduleFolder, "projectStatus.rds"))
```

Comment:
There could be a weird race condition here if a machine is in a multithreaded environment and a different states of installing. That's probably a rare edge case though - just bare this in mind if we're ever clustering `Strategus` different workers could have independent processes that call `verifyModuleInstallation` so we may want to add a lock file.

Reply:
Agreed - this may be worth noting for a future version of Strategus.